### PR TITLE
visualize the A* plan making process with graphviz

### PR DESCRIPTION
**Pull request by TMPxyz**
This changelist added a T4 template (PlanDebugger.tt) and modified the AStar.cs to allow output of the A* planning process into log files, which can be used by Graphviz to generate a picture showing the plan making process.

Only the agent with _agent.debugPlan = true_ would trigger this logging behaviour;

Some extra modifications to the interfaces are added to extract the info needed by the logs;

The generated picture would look like the example below.

debug

The picture might become huge when the actions increase, it might be a useful improvement to add switches to turn off some info output, hence reduce the final picture size.

### CODE REVIEW

File: ReGoap/Planner/AStar.cs [hunk 1]
```diff
@@ -28,8 +28,47 @@ void ClearNodes()
             createdNodes.Clear();
         }
 
-        public INode<T> Run(INode<T> start, T goal, int maxIterations = 100, bool earlyExit = true, bool clearNodes = true)
+        private bool _debugPlan = false;
```
Comment (luxkun) [inline ReGoap/Planner/AStar.cs]: Can you follow the style of the rest of the code, by removing the starting _?

File: ReGoap/Planner/AStar.cs [hunk 2]
```diff
@@ -28,8 +28,47 @@ void ClearNodes()
...
+        private Assets.ReGoap.Planner.PlanDebugger _debugger;
+
+        private void _DebugPlan(INode<T> node, INode<T> parent)
```
Comment (luxkun) [inline ReGoap/Planner/AStar.cs]: Samething here

File: ReGoap/Planner/AStar.cs [hunk 3]
```diff
@@ -28,8 +28,47 @@ void ClearNodes()
             createdNodes.Clear();
         }
 
-        public INode<T> Run(INode<T> start, T goal, int maxIterations = 100, bool earlyExit = true, bool clearNodes = true)
+        private bool _debugPlan = false;
+        private Assets.ReGoap.Planner.PlanDebugger _debugger;
```
Comment (luxkun) [inline ReGoap/Planner/AStar.cs]: Samething here

File: ReGoap/Planner/AStar.cs [hunk 4]
```diff
@@ -40,16 +79,23 @@ public INode<T> Run(INode<T> start, T goal, int maxIterations = 100, bool earlyE
...
 
             frontier.Enqueue(start, start.GetCost());
+
+            _DebugPlan(start, null);
+
             var iterations = 0;
             while ((frontier.Count > 0) && (iterations < maxIterations) && (frontier.Count + 1 < frontier.MaxSize))
             {
                 var node = frontier.Dequeue();
+                //Utilities.ReGoapLogger.Log(string.Format("\n++++Explored action: {0}({3}), state ({1})\n goal ({2})\n effect: ({4})", node.Name, node.GetState(), node.GoalString, node.GetCost(), node.EffectString));
```
Comment (luxkun) [inline ReGoap/Planner/AStar.cs]: Remove

File: ReGoap/Planner/AStar.cs [hunk 5]
```diff
@@ -75,11 +122,16 @@ public INode<T> Run(INode<T> start, T goal, int maxIterations = 100, bool earlyE
                         else
                             break;
                     }
+
+                    _DebugPlan(child, node);
+
+                    //Utilities.ReGoapLogger.Log(string.Format("    Enqueue frontier: {0}, cost: {1}", child.Name, childCost));
```
Comment (luxkun) [inline ReGoap/Planner/AStar.cs]: Remove

File: ReGoap/Planner/AStar.cs [hunk 6]
```diff
@@ -94,6 +146,11 @@ public interface INode<T>
         float GetPathCost();
         INode<T> GetParent();
         bool IsGoal(T goal);
+        //int AstarID { get; } //used for planDebug
```
Comment (luxkun) [inline ReGoap/Planner/AStar.cs]: Remove

File: ReGoap/Planner/PlanDebugger1.cs [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+    using System.Linq;
+    using System.Text;
+    using System.Collections.Generic;
+    using System;
+    
+    /// <summary>
+    /// Class to produce the template output
+    /// </summary>
+    
+    #line 1 "E:\works\programs\GameProjects\TestGround\TestReGOAP\Assets\ReGoap\Planner\PlanDebugger.tt"
```
Comment (luxkun) [inline ReGoap/Planner/PlanDebugger1.cs]: What's this file? why is there a full path of your hard drive here?
Also why is it named PlanDebugger**1**?

File: ReGoap/Planner/ReGoapPlanner.cs [hunk 8]
```diff
@@ -70,7 +70,7 @@ public ReGoapPlanner(ReGoapPlannerSettings settings = null)
 
                 goalState = goalState.Clone();
                 var leaf = (ReGoapNode<T, W>)astar.Run(
-                    ReGoapNode<T, W>.Instantiate(this, goalState, null, null), goalState, settings.MaxIterations, settings.PlanningEarlyExit);
+                    ReGoapNode<T, W>.Instantiate(this, goalState, null, null), goalState, settings.MaxIterations, settings.PlanningEarlyExit, debugPlan : agent.debugPlan);
```
Comment (luxkun) [inline ReGoap/Planner/ReGoapPlanner.cs]: I would put the debugPlan inside the Planner settings (where MaxIterations is set) and call it something more sensible, ex. "graphvizDebug"

File: ReGoap/Unity/ReGoapAgent.cs [hunk 9]
```diff
@@ -15,6 +15,8 @@ public class ReGoapAgent<T, W> : MonoBehaviour, IReGoapAgent<T, W>, IReGoapAgent
 
         public bool CalculateNewGoalOnStart = true;
 
+        public bool shouldDebugPlan = false;
```
Comment (luxkun) [inline ReGoap/Unity/ReGoapAgent.cs]: As Written before, this should probably go in the Planner settings

File: ReGoap/Planner/PlanDebugger1.cs (hunk 7 above)
Comment (TMPxyz) [reply ReGoap/Planner/PlanDebugger1.cs]: PlanDebugger.tt is a [T4](https://msdn.microsoft.com/en-us/library/ee844259.aspx) code generation file, which is used by VisualStudio to generate this PlanDebugger1.cs.

PlanDebugger1.cs is used to create the logs for graphviz;

File: ReGoap/Unity/ReGoapAgent.cs (hunk 9 above)
Comment (TMPxyz) [reply ReGoap/Unity/ReGoapAgent.cs]: Got it, will make changes according to the reviews;

**Comment by luxkun** [comment]
Hey TMPxyz, I'll be doing these changes myself on another branch, thanks for your pull request! 😄

**Pull request merged**